File: app/pix.py
"""
Geracao de Pix estatico no padrao EMV (BR Code), com CRC16-CCITT.
Preparado para no futuro ser substituido por integracao com API
(Mercado Pago, Asaas, Gerencianet, PagSeguro) sem mudar a interface
publica: build_pix_payload() / generate_qr_image().
"""
import re
import unicodedata
import io

import qrcode


def _emv_field(field_id: str, value: str) -> str:
    length = f"{len(value):02d}"
    return f"{field_id}{length}{value}"


def _sanitize(text: str, max_len: int) -> str:
    if not text:
        text = ""
    normalized = unicodedata.normalize("NFKD", text).encode("ascii", "ignore").decode("ascii")
    normalized = re.sub(r"[^A-Za-z0-9 ]", "", normalized).strip().upper()
    return normalized[:max_len] if normalized else "NAO INFORMADO"


def _crc16_ccitt(payload: str) -> str:
    """CRC16-CCITT (falso), poly 0x1021, init 0xFFFF - padrao usado pelo Bacen no Pix."""
    poly = 0x1021
    crc = 0xFFFF
    data = payload.encode("utf-8")
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ poly) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return f"{crc:04X}"


def build_pix_payload(pix_key: str, merchant_name: str, merchant_city: str,
                       amount: float, description: str = "", txid: str = "***") -> str:
    """Monta o BR Code (copia-e-cola) do Pix estatico, com valor fixo e CRC16 no final."""
    if not pix_key:
        raise ValueError("Chave Pix nao configurada. Configure em Configuracoes > Pix.")

    merchant_account_info = (
        _emv_field("00", "BR.GOV.BCB.PIX")
        + _emv_field("01", pix_key)
        + (_emv_field("02", description[:40]) if description else "")
    )

    txid_clean = re.sub(r"[^A-Za-z0-9]", "", txid or "***") or "***"

    payload = (
        _emv_field("00", "01")
        + _emv_field("01", "12")
        + _emv_field("26", merchant_account_info)
        + _emv_field("52", "0000")
        + _emv_field("53", "986")
        + _emv_field("54", f"{amount:.2f}")
        + _emv_field("58", "BR")
        + _emv_field("59", _sanitize(merchant_name, 25))
        + _emv_field("60", _sanitize(merchant_city, 15))
        + _emv_field("62", _emv_field("05", txid_clean[:25]))
    )

    payload_for_crc = payload + "6304"
    crc = _crc16_ccitt(payload_for_crc)
    return payload_for_crc + crc


def generate_qr_image(payload: str):
    """Retorna um objeto PIL.Image com o QR Code do payload Pix."""
    qr = qrcode.QRCode(
        version=None,
        error_correction=qrcode.constants.ERROR_CORRECT_M,
        box_size=8,
        border=3,
    )
    qr.add_data(payload)
    qr.make(fit=True)
    return qr.make_image(fill_color="black", back_color="white").convert("RGB")


def generate_qr_bytes(payload: str) -> bytes:
    img = generate_qr_image(payload)
    buf = io.BytesIO()
    img.save(buf, format="PNG")
    return buf.getvalue()

File: cloud-pwa/src/lib/pix.js
import QRCode from "qrcode";

function emvField(id, value) {
  const text = String(value ?? "");
  return `${id}${String(text.length).padStart(2, "0")}${text}`;
}

function sanitize(text, maxLen) {
  const clean = String(text || "")
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^A-Za-z0-9 ]/g, "")
    .trim()
    .toUpperCase();
  return (clean || "NAO INFORMADO").slice(0, maxLen);
}

function crc16(payload) {
  let crc = 0xffff;
  for (let i = 0; i < payload.length; i += 1) {
    crc ^= payload.charCodeAt(i) << 8;
    for (let bit = 0; bit < 8; bit += 1) {
      crc = crc & 0x8000 ? ((crc << 1) ^ 0x1021) & 0xffff : (crc << 1) & 0xffff;
    }
  }
  return crc.toString(16).toUpperCase().padStart(4, "0");
}

export function buildPixPayload({
  pixKey,
  merchantName,
  merchantCity,
  amount,
  description = "",
  txid = "***",
}) {
  if (!pixKey) throw new Error("Chave Pix nao configurada.");
  const merchantAccount =
    emvField("00", "BR.GOV.BCB.PIX") +
    emvField("01", pixKey) +
    (description ? emvField("02", String(description).slice(0, 40)) : "");
  const cleanTxid = String(txid || "***").replace(/[^A-Za-z0-9]/g, "") || "***";
  const payload =
    emvField("00", "01") +
    emvField("01", "12") +
    emvField("26", merchantAccount) +
    emvField("52", "0000") +
    emvField("53", "986") +
    emvField("54", Number(amount || 0).toFixed(2)) +
    emvField("58", "BR") +
    emvField("59", sanitize(merchantName, 25)) +
    emvField("60", sanitize(merchantCity, 15)) +
    emvField("62", emvField("05", cleanTxid.slice(0, 25)));
  const withCrcMarker = `${payload}6304`;
  return `${withCrcMarker}${crc16(withCrcMarker)}`;
}

export async function pixQrDataUrl(payload) {
  return QRCode.toDataURL(payload, { margin: 2, width: 260 });
}
